kernel/src/engine/arrow_utils.rs

```diff
@@ -1,89 +1,1338 @@
+            // if our children are not ordered somehow, we need a transform
+            ReorderIndexTransform::Child(ref children) => !is_ordered(children),
+            // no transform needed
+            ReorderIndexTransform::None => false,
+        }
+    }
+}
+
+// count the number of physical columns, including nested ones in an `ArrowField`
+fn count_cols(field: &ArrowField) -> usize {
+    _count_cols(field.data_type())
+}
+
+fn _count_cols(dt: &ArrowDataType) -> usize {
+    match dt {
+        ArrowDataType::Struct(fields) => fields.iter().map(|f| count_cols(f)).sum(),
+        ArrowDataType::Union(fields, _) => fields.iter().map(|(_, f)| count_cols(f)).sum(),
+        ArrowDataType::List(field)
+        | ArrowDataType::LargeList(field)
+        | ArrowDataType::FixedSizeList(field, _)
+        | ArrowDataType::Map(field, _) => count_cols(field),
+        ArrowDataType::Dictionary(_, value_field) => _count_cols(value_field.as_ref()),
+        _ => 1, // other types are "real" fields, so count
+    }
+}
+
+/// helper function, does the same as `get_requested_indices` but at an offset. used to recurse into
+/// structs, lists, and maps. `parquet_offset` is how many parquet fields exist before processing
+/// this potentially nested schema. returns the number of parquet fields in `fields` (regardless of
+/// if they are selected or not) and reordering information for the requested fields.
+fn get_indices(
+    start_parquet_offset: usize,
+    requested_schema: &Schema,
+    fields: &Fields,
+    mask_indices: &mut Vec<usize>,
+) -> DeltaResult<(usize, Vec<ReorderIndex>)> {
+    let mut found_fields = HashSet::with_capacity(requested_schema.fields.len());
+    let mut reorder_indices = Vec::with_capacity(requested_schema.fields.len());
+    let mut parquet_offset = start_parquet_offset;
+    for (parquet_index, field) in fields.iter().enumerate() {
+        debug!(
+            "Getting indices for field {} with offset {parquet_offset}, with index {parquet_index}",
+            field.name()
+        );
+        match field.data_type() {
+            ArrowDataType::Struct(fields) => {
+                if let Some((index, _, requested_field)) =
+                    requested_schema.fields.get_full(field.name())
+                {
+                    if let DataType::Struct(ref requested_schema) = requested_field.data_type {
+                        let (parquet_advance, children) = get_indices(
+                            parquet_index + parquet_offset,
+                            requested_schema.as_ref(),
+                            fields,
+                            mask_indices,
+                        )?;
+                        // advance the number of parquet fields, but subtract 1 because the
+                        // struct will be counted by the `enumerate` call but doesn't count as
+                        // an actual index.
+                        parquet_offset += parquet_advance - 1;
```

Comment:
Just to verify: This also works for deeper schemas because each level of nesting introduces a new `Struct` which will also decrement the count appropriately?

Reply:
yep, exactly